File: sdk/csharp/src/Providers/Tempmailc.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// TempMailC 渠道（tempmailc.com/api/v1）。
/// 创建：GET /new；收信：GET /inbox?email=，逐封 GET /message 补全详情（失败回退列表项）。
/// </summary>
public static class Tempmailc
{
    private const string ApiBase = "https://tempmailc.com/api/v1";

    private static Dictionary<string, string> Headers() => new() { ["Accept"] = "application/json" };

    private static Dictionary<string, object?> FlattenListMessage(JsonObject raw, string recipient) => new()
    {
        ["id"] = Json.Str(raw, "id"),
        ["from"] = Json.Str(raw, "from"),
        ["to"] = recipient,
        ["subject"] = Json.Str(raw, "subject"),
        ["timestamp"] = raw["ts"] is JsonValue tv && tv.TryGetValue<long>(out var ts) ? ts : (object?)Json.Str(raw, "ts"),
        ["read"] = (raw["read"] as JsonValue)?.GetValue<bool>() ?? false,
    };

    private static JsonObject? FetchMessage(string email, string messageId)
    {
        var resp = Http.Get($"{ApiBase}/message?email={WebUtility.UrlEncode(email)}&msg_id={WebUtility.UrlEncode(messageId)}", Headers());
        if (!resp.Ok) return null;
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || (data["ok"] as JsonValue)?.GetValue<bool>() == false) return null;
        return data;
    }

    /// <summary>创建邮箱</summary>
    public static EmailInfo Generate()
    {
        var resp = Http.Get($"{ApiBase}/new", Headers());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        var email = data is null ? "" : Json.Str(data, "email").Trim();
        if (data is null || (data["ok"] as JsonValue)?.GetValue<bool>() != true || email.Length == 0 || !email.Contains('@'))
            throw new Exception("tempmailc: invalid new email response");
        return new EmailInfo("tempmailc", email);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email)
    {
        var address = (email ?? "").Trim();
        if (address.Length == 0) throw new Exception("tempmailc: empty email");
        var resp = Http.Get($"{ApiBase}/inbox?email={WebUtility.UrlEncode(address)}", Headers());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || (data["ok"] as JsonValue)?.GetValue<bool>() != true)
            throw new Exception("tempmailc: inbox response failed");
        var rows = data["messages"] as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var item in rows)
        {
            if (item is not JsonObject ro) continue;
            var messageId = Json.Str(ro, "id").Trim();
            var detail = messageId.Length > 0 ? FetchMessage(address, messageId) : null;
            var flat = detail is not null ? Json.ToDict(detail) : FlattenListMessage(ro, address);
            result.Add(Normalize.NormalizeEmail(flat, address));
        }
        return result;
    }
}

File: sdk/php/src/Providers/Tempmailc.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * TempMailC 渠道实现 — https://tempmailc.com
 *
 * GET /new 创建邮箱；GET /inbox?email= 列表；GET /message?email=&msg_id= 详情。
 */
final class Tempmailc
{
    private const CHANNEL = 'tempmailc';
    private const API_BASE = 'https://tempmailc.com/api/v1';

    /** @var array<string,string> */
    private const HEADERS = ['Accept' => 'application/json'];

    public static function generate(): EmailInfo
    {
        $resp = HttpClient::get(self::API_BASE . '/new', self::HEADERS);
        $data = HttpClient::json($resp);
        $email = trim((string) ($data['email'] ?? ''));
        if (empty($data['ok']) || $email === '' || !str_contains($email, '@')) {
            throw new \RuntimeException('tempmailc: 无效的创建响应');
        }
        return new EmailInfo(self::CHANNEL, $email);
    }

    /**
     * 拉取单封邮件详情，失败返回 null。
     *
     * @return array<mixed>|null
     */
    private static function fetchMessage(string $email, string $messageId): ?array
    {
        $resp = HttpClient::get(self::API_BASE . '/message', self::HEADERS, query: [
            'email' => $email,
            'msg_id' => $messageId,
        ]);
        if ($resp->getStatusCode() >= 400) {
            return null;
        }
        $data = HttpClient::json($resp);
        if (empty($data) || ($data['ok'] ?? null) === false) {
            return null;
        }
        return $data;
    }

    /**
     * @param array<mixed> $raw
     * @return array<mixed>
     */
    private static function flattenList(array $raw, string $recipient): array
    {
        return [
            'id' => $raw['id'] ?? '',
            'from' => $raw['from'] ?? '',
            'to' => $recipient,
            'subject' => $raw['subject'] ?? '',
            'timestamp' => $raw['ts'] ?? null,
            'read' => $raw['read'] ?? null,
        ];
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token = null): array
    {
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('tempmailc: 邮箱地址为空');
        }
        $resp = HttpClient::get(self::API_BASE . '/inbox', self::HEADERS, query: ['email' => $addr]);
        $data = HttpClient::json($resp);
        if (empty($data['ok'])) {
            throw new \RuntimeException('tempmailc: inbox 响应失败');
        }
        $rows = $data['messages'] ?? [];
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $item) {
            if (!is_array($item)) {
                continue;
            }
            $messageId = trim((string) ($item['id'] ?? ''));
            $detail = $messageId !== '' ? self::fetchMessage($addr, $messageId) : null;
            $out[] = Normalize::email($detail ?? self::flattenList($item, $addr), $addr);
        }
        return $out;
    }
}
